docs/content/configuration/index.md

```diff
@@ -1319,6 +1321,8 @@ You can optionally only configure caching to be enabled on the Historical by set
 |`druid.historical.cache.useCache`|true, false|Enable the cache on the Historical.|false|
 |`druid.historical.cache.populateCache`|true, false|Populate the cache on the Historical.|false|
 |`druid.historical.cache.unCacheable`|All druid query types|All query types to not cache.|["groupBy", "select"]|
+|`druid.historical.cache.numBackgroundThreads`|If greater than 0, cache will be populated in the background thread pool of the configured size|0|
```

Comment:
I thought there was some reason that background cache population wasn't documented. IIRC it was broken somehow (something about generating corrupt / incorrect cached data). But unfortunately I can't remember what was wrong or if it has been fixed yet.

If it turns out that the feature _does_ work now, I believe it still does have one serious flaw — no load shedding means it can lead to your JVM running OOM.

IMO, if we think the feature is still broken (in the sense that it corrupts caches), let's _not_ document it.

If we think the feature works but has this potential OOM issue, then I think it's OK to document it, but the docs should call out the potential for OOMs due to lack of load shedding.

Reply:
I did a bit of searching and came across some stuff, but nothing conclusive. [This comment](https://github.com/apache/incubator-druid/issues/3313#issuecomment-236914678), in an issue about missing documentation, indicates what you are suggesting. But there also exists #3943 which purports to fix thread safety issues with the background cache and is more recent than the discussion in that issue. Issue #4199, which I found linked to #3943 and indicates that the problem wasn't visible with cache disabled, but ultimately seemed related to aggregator thread safety since it was marked resolved by #4304, so whether or not this is broken might tie into that? (So #3956, #4304, https://github.com/apache/incubator-druid/pull/5148#discussion_r170906998, proposal #8031, and many other related discussions.)

Maybe if we document the setting as experimental and mention the potential for oom it would be ok to document? That would be my preference over leaving it undocumented unless it is truly broken always, but I haven't been able to fully determine that yet.

Reply:
Due to the controversy about how to word the doc for background cache population, I'd suggest breaking that out into its own PR and leaving this one to only include the lock-removal fix for foreground cache population. These two things are only tangentially related. In the other PR, we could discuss:

- Whether or not we are going to document background caching at this time (given the above thread).
- What exactly the docs should say.